### ISSUE TITLE
core 3.0 preview 6 broke COM Interop usage, compiles but raises System.TypeLoadException "The signature is incorrect" on runtime

### ISSUE THREAD

**Issue by dude202** [dotnet/coreclr#25651]
Core 3.0 test project references and uses COM component.
Preview 5 does function without problems. Proven by uninstalling Preview 6 and installing Preview 5 without any Project/Code changes.

**Comment by dude202** [comment]
Repro is actually non-trivial. It is about referencing and calling 32bit COM dll packed in COM+ app from a 64bit C# library (more precise anyCPU on x64 machine) using following to create an (possibly remote) object:
var server = (ServerLib.Server)Activator.CreateInstance(Type.GetTypeFromProgID("<ProgID>", "<machine name>", true));
Once again, till preview 6 everything worked fine, no problems whatsoever.

The following works also in preview 6:
var server = Activator.CreateInstance(Type.GetTypeFromProgID("<ProgID>", "<machine name>", true));
So I conclude that type casting is the problem and triggers "The signature is incorrect" exception.

**Comment by AaronRobinsonMSFT** [comment]
>var server = (ServerLib.Server)Activator.CreateInstance(Type.GetTypeFromProgID("", "", true));

@dude202 Is `ServerLib.Server` the actual coclass? If so, this support was explicitly removed. Instead casting to an actual interface is what is required. The "signature is incorrect" exception doesn't make sense if this is a coclass scenario since it should be "invalid cast", so I am thinking this is probably something else. The 32/64 mix through a remote COM proxy is interesting and I will look into that today.

**Comment by AaronRobinsonMSFT** [comment]
>using following to create an (possibly remote) object:

@dude202 Just so I can clarify, does your usage of "remote" here indicate another machine? I ask because if the server is 32-bit and the client is running 64-bit the server will definitely be out-of-proc regardless. From the CLR perspective "remote" is out-of-proc whether it is on the current machine or on another machine.

**Comment by dude202** [comment]
Yes, it is out-of-proc anyway, remote or not. Thanks for the effort, please dont give up on this scenario, because:
1. It already worked
2. It is the only way (I know of) to reuse old 32bit COM components in newest technology context (plus the bitness switch).

**Comment by AaronRobinsonMSFT** [comment]
@dude202 Definitely not giving up on these kind of scenarios. The fact that it already worked is something we need to ensure was either regressed for a good reason or is a bug. I say "regressed for a good reason" in the sense we may have broken it in order to make a more common scenario better. However, the desire to reuse 32-bit COM components in a 64-bit scenario is something we definitely want to ensure works in .NET Core, it may just need to be done in a slightly different way from the client perspective.

A few more questions so I can ensure I understand the scenario:

1) Is the server managed or native?
    - If managed, is this written against .NET Framework or .NET Core?
1) What are the server's registration settings?
    - How was it registered? `RegAsm.exe`? `regsvr32.exe`? `.reg` files?
    - Apartment type?
    - Registered with a TypeLib?
    - COM Proxy/stubs?
1) How is the `ServerLib.Server` type defined in managed?

Have you tried the scenario in preview 7?

**Comment by dude202** [comment]
I have not tried preview 7, is it available? As daily build or...? 
Server is a vb6 dll installed as COM+ application, that means STA, native, registered with regsvr32 with default proxy/stubs. Referenced using Interop that VS generates on-the-fly from dll, no typelib import.
Managed caller is strong named assembly, but the same error happens also without it.

**Comment by dude202** [comment]
ServerLib.Server is the default interface of the coclass (its coclass is named "ServerClass" in VS Interop, "Server" is inherited from "_Server")

**Comment by AaronRobinsonMSFT** [comment]
@dude202 VB6... Okay well how about this. Can you provide a simple VB6 compiled server that I can use? If you can share that out it would help me to root cause the exact issue.

**Comment by dude202** [comment]
[VB6TestServer.zip](https://github.com/dotnet/coreclr/files/3387946/VB6TestServer.zip)
[TestCOMCore.zip](https://github.com/dotnet/coreclr/files/3387947/TestCOMCore.zip)
COM+ app (to import) and c# core 3 project to trigger the exception.

**Comment by AaronRobinsonMSFT** [comment]
@dude202 I was able to run the following using the supplied VB6 asset. Note that I didn't install the MSI - sorry, we don't do that. I manually registered the component - should be the same as `regsvr32.exe`. In order to consume the 32-bit component from the 64-bit .NET Core I used the [DllSurrogate](https://docs.microsoft.com/windows/win32/com/dllsurrogate) mechanism.

```CSharp
[STAThread]
static void Main(string[] args)
{
    var type = Type.GetTypeFromProgID("ServerLib.Server", null, true);
    Console.WriteLine($"{type}");

    var server = (ServerLib.Server)Activator.CreateInstance(type);
    Console.WriteLine($"{server}");

    server.TestMethod();
}
```

The registration for DllSurrogate was done as follows:
```
[HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Classes\CLSID\{E07693C8-A5BB-4080-8F8A-B6375F56C927}]
@="ServerLib.Server"
"AppID"="{DB87B5FD-8A2A-410A-A755-6DE450A024A7}"

[HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Classes\AppID\{DB87B5FD-8A2A-410A-A755-6DE450A024A7}]
"DllSurrogate"=""
```

**Comment by dude202** [comment]
Well, your console test (and dll registration) leaves me with a strange problem:
An unhandled exception of type 'System.IO.FileNotFoundException' occurred in Unknown Module.
Could not load file or assembly 'Interop.ServerLib, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
before any managed code is executed (no call stack).
I get this error independent of "surrogate" or "plain" com dll registration (which, I know, cannot function for in-process activation, I tried it just for testing).
Anyway, even if it had worked, the scenario with DllSurrogate is not the one I can use. vb6 dll is packed in a COM+ application, it runs already out-of-process (and will also need a remote activation).
My test solution with console that calls test assembly that calls ServerLib dll in com+ app does not have “could not load” problem, and I think I found out when it triggers “The signature is incorrect” error and when not: class “Server” implements COM+ Services (comsvcs.dll) interface “ObjectControl” for better integration in com+ context. If it does, “The signature is incorrect” triggers (I tried to add a reference to comsvcs in test assembly manually, it makes no difference). If I remove interface implementation from the “Server” test class, my managed test runs without error. Unfortunately, the real vb6 server class implements the interface and can’t be changed.

**Comment by AaronRobinsonMSFT** [comment]
>An unhandled exception of type 'System.IO.FileNotFoundException' occurred in Unknown Module.
Could not load file or assembly 'Interop.ServerLib, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
before any managed code is executed (no call stack).

This can be addressed by providing the interop assembly with the console application or embedding the types. Set `<EmbedInteropTypes>True</EmbedInteropTypes>` in the `ComReference` for `ServerLib`.

>vb6 dll is packed in a COM+ application, it runs already out-of-process (and will also need a remote activation).

Ah. That is a key aspect of this scenario - thanks.

>class “Server” implements COM+ Services (comsvcs.dll) interface “ObjectControl” for better integration in com+ context

This should be addressed by embedding the types in the client assembly - see above.

**Comment by dude202** [comment]
Lo and behold... That's really the solution for preview 6! "Not embedding" somehow works in preview 5? I could swear I tried that already... Well, thank you very much for your time and effort, I look forward to further testing core 3 previews.

**Comment by AaronRobinsonMSFT** [comment]
@dude202 Yay! Glad to hear it. I will leave this issue since the issue is that non-embedding seems to have regressed in this scenario.
